compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Monoid.hs

```diff
@@ -0,0 +1,68 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE OverloadedStrings, DerivingStrategies #-}
+
+-- | Monoid with which to render documentation.
+module DA.Daml.Doc.Render.Monoid
+  ( module DA.Daml.Doc.Render.Monoid
+  ) where
+
+import DA.Daml.Doc.Types
+import qualified Data.Set as Set
+import qualified Data.Text as T
+
+-- | Environment in which to generate final documentation.
+newtype RenderEnv = RenderEnv (Set.Set Anchor)
+    deriving newtype (Semigroup, Monoid)
+
+-- | Is the anchor available in the rendering environment? Renderers should avoid
+-- generating links to anchors that don't actually exist.
+--
+-- One reason an anchor may be unavailable is because of a @-- | HIDE@ directive.
+-- Another possibly reason is that the anchor refers to a definition in another
+-- package (and at the moment it's not possible to link accross packages).
+renderAnchorAvailable :: RenderEnv -> Anchor -> Bool
+renderAnchorAvailable (RenderEnv anchors) anchor = Set.member anchor anchors
+
+-- | Renderer output. This is the set of anchors that were generated, and a
+-- list of output functions that depend on that set. The goal is to prevent
+-- the creation of spurious anchors links (i.e. links to anchors that don't
+-- exist).
+--
+-- (In theory this could be done in two steps, but that seems more error prone
+-- than building up both steps at the same time, and combining them at the
+-- end, as is done here.)
+--
+-- Using a newtype here so we can derive the semigroup / monoid instances we
+-- want automatically. :-)
+newtype RenderOut = RenderOut (RenderEnv, [RenderEnv -> [T.Text]])
```

Comment:
Should we use a `DList` here? The `Semigroup` instance of `[a]` has its performance issues.

Reply:
That's a pretty good idea :)

Reply:
^ I'll change it in a future PR, when the renderer thing becomes more settled. (Still have to figure out how to render to multiple files.)